stan/math/fwd/fun/unit_vector_constrain.hpp

```diff
@@ -15,42 +15,37 @@
 namespace stan {
 namespace math {
 
-template <typename T, int R, int C>
-inline Eigen::Matrix<fvar<T>, R, C> unit_vector_constrain(
-    const Eigen::Matrix<fvar<T>, R, C>& y) {
-  using Eigen::Matrix;
-  using std::sqrt;
-
-  Matrix<T, R, C> y_t(y.size());
-  for (int k = 0; k < y.size(); ++k) {
-    y_t.coeffRef(k) = y.coeff(k).val_;
-  }
-
-  Matrix<T, R, C> unit_vector_y_t = unit_vector_constrain(y_t);
-  Matrix<fvar<T>, R, C> unit_vector_y(y.size());
-  for (int k = 0; k < y.size(); ++k) {
-    unit_vector_y.coeffRef(k).val_ = unit_vector_y_t.coeff(k);
-  }
-
-  T squared_norm = dot_self(y_t);
-  T norm = sqrt(squared_norm);
-  T inv_norm = inv(norm);
-  Matrix<T, Eigen::Dynamic, Eigen::Dynamic> J
-      = divide(tcrossprod(y_t), -norm * squared_norm);
-
-  for (int m = 0; m < y.size(); ++m) {
+template <typename EigMat, require_eigen_vt<is_fvar, EigMat>* = nullptr>
+inline auto unit_vector_constrain(const EigMat& y) {
+  using eig_mat = std::decay_t<EigMat>;
+  using eigen_type = typename eig_mat::PlainObject;
```

Comment:
```suggestion
  using T_plain = typename eig_mat::PlainObject;
```

I find the name `eigen_type` a bit confusing.

Reply:
We have much better type traits now so I'll clean this whole thing up